### PR TITLE
[Woo POS] Initial implementation of payment cancellation

### PR DESCRIPTION
**Pull request by jaclync**
Part of #12869 

## Description

When leaving the totals screen which can be in any state (reader connection or payment collection), we want to cancel any relevant task. The merchant can leave the totals screen for:
- Tap `Add More` to change items in the cart
- Tap `Exit POS` to leave the POS mode
- The merchant can also force quit the app - it should be handled the same as in order form IPP

## Testing information

Test that after leaving the totals view to change the cart for each of the following states, payment works on the updated cart (due to the pending [payment intent task](https://github.com/orgs/woocommerce/projects/247/views/6?sliceBy%5Bvalue%5D=iOS+%F0%9F%8D%8E&pane=issue&itemId=67688689), please make sure the updated cart has a lower amount than the first time payment is ready):

### Critical flow

- scanningForReaders(endSearch: () -> Void)
  - [x] With a reader active nearby, the reader isn't auto-connected after leaving the totals screen to the product selector
- connectingToReader
  - [x] The reader still gets connected, but fails with invalidated payment in payment collection. The auto-payment succeeds after updating the cart
- preparingForPayment(cancelPayment: () -> Void)
  - [x] Success canceling PI, and subsequent payment succeeds
- foundReader(name: String, connect: () -> Void, continueSearch: () -> Void, endSearch: () -> Void)
  - [x] Reader isn't connected after leaving the totals screen to the product selector
- tapSwipeOrInsertCard(inputMethods: CardReaderInput, cancelPayment: () -> Void)
  - [x] After leaving the totals screen in this state and then back, the reader can still transition to the ready state
- paymentSuccess(done: () -> Void)
  - [x] Error is shown about the order already paid. This should be made impossible when we make the totals UI fullscreen
- paymentError(error: any Error, tryAgain: () -> Void, cancelPayment: () -> Void)
  - [x] Tested this case by switching device to offline after the ready state then tapping the card to pay, with error `Hardware.CardReaderServiceError.paymentCapture(underlyingError: Hardware.UnderlyingError.notConnectedToInternet)`. After going online and then retrying/canceling, subsequent payment succeeds
- paymentErrorNonRetryable(error: any Error, cancelPayment: () -> Void)
  - [x] Using steps to repro in https://github.com/woocommerce/woocommerce-ios/pull/13100#issuecomment-2186860471
- processing
  - [x] Depending on the time cancellation is triggered, PI could be canceled successfully or not (we should disable any navigation away from this screen https://github.com/woocommerce/woocommerce-ios/issues/13115)
- displayReaderMessage(message: String)
  - [x] When the message is `Remove Card`, PI successfully canceled and subsequent payment succeeds
- validatingOrder(cancelPayment: () -> Void)
  - [x] After leaving the totals screen in this state and then back, the reader can still transition to the ready state and be paid successfully

### Test with a simulated reader

- showReaderList when multiple readers are found
  - I couldn't repro this
- updateProgress(requiredUpdate: Bool, progress: Float, cancelUpdate: (() -> Void)?)
  - [x] Reader discovery canceled 
- updateFailed(tryAgain: () -> Void, cancelUpdate: () -> Void)
- updateFailedNonRetryable(cancelUpdate: () -> Void)
- updateFailedLowBattery(batteryLevel: Double?, cancelUpdate: () -> Void)

### Unhappy cases

- scanningFailed(error: Error, endSearch: () -> Void)
  - I couldn't repro this
- bluetoothRequired(error: Error, endSearch: () -> Void)
  - Tested without a simulated reader in iPad simulator, but couldn't test the recovery flow
- connectingFailed(error: Error, retrySearch: () -> Void, endSearch: () -> Void)
- connectingFailedNonRetryable(error: Error, endSearch: () -> Void)
- connectingFailedUpdatePostalCode(retrySearch: () -> Void, endSearch: () -> Void)
- connectingFailedChargeReader(retrySearch: () -> Void, endSearch: () -> Void)
- connectingFailedUpdateAddress(wcSettingsAdminURL: URL, retrySearch: () -> Void, endSearch: () -> Void)
- cancelledOnReader

### Not supported

- selectSearchType(tapToPay: () -> Void, bluetooth: () -> Void, endSearch: () -> Void)

## Screenshots

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (joshheald) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: Good point, but FYI this is about having Apple Pay (wallet) on the device, and not  wanting the merchant's Apple Pay card to be accidentally used for payment if they stick a card reader to the back of the device.

Not needed for an iPad, but on an iPhone it is. iPads don't have the wallet capability.

This is definitely the sort of thing we might miss if we don't implement it now though!

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
Comment (joshheald) [inline WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:210]: We'll need this regardless when Bozidar's changes are merged, but did you find that the cancellation wasn't called automatically because of the `task` modifier we used? I thought it would be.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: I don't see a clear way to cancel the connection easily from the [Stripe doc](https://docs.stripe.com/terminal/payments/connect-reader?reader-type=bluetooth). Checked the `BluetoothReaderDelegate` as well.

From testing, leaving the totals screen while the state is `connectingToReader` sometimes results in the reader not being able to connect afterward (same in the Menu > Payments > Manage Card Reader) and sometimes the connection continues.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: From testing, at the time of `cancelPayment` (canceling Payment Intent (PI)) at this state, the PI could already be created and thus an error was thrown from canceling the PI. This also results in the payment state still advancing to the `tapSwipeOrInsertCard` state. We need to make sure the PI is recreated (since updating PI isn't an option pdfdoF-54W-p2) when the cart is updated, after the order sync integration is merged.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (joshheald) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: I think we should let the connection continue here, but prevent the payment. Implemented an approach to this in 5f511b9, though I'm not sure how good it is. Let me know what you think.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (joshheald) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: We could potentially move the `invalidated` check to right before `attemptPayment` is called in the use case, which might improve this...

I suspect it's that the payment has not actually begun when we call `cancelPayment`, so it does nothing, but the order check succeeds and _then_ the payment starts.

**Review by joshheald — COMMENTED** [review]
I've iterated on this a bit but I'm still testing the various situations.

I'd like cancellation on the service to be as effective as possible regardless of where it's called from... but we will have protection from some cases in POS by preventing people from going back after the card is tapped.

The `.processing` case is particularly difficult, and that really just reflects that we can't always effectively cancel an in-progress payment. What do you think the right behaviour is in this case? We can't throw anything from the task's cancellation handler, unfortunately.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (joshheald) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: Another way would be to provide a specialised PaymentOrchestrator, which could have the invalidation behaviour in it...

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift:94]: Love it! Nice that it's called for all cases, for one particular payment collection task 💯

**Comment by jaclync** [comment]
Just merged [feat/12763-pos-order-totals-order-creation-sync](https://github.com/woocommerce/woocommerce-ios/tree/feat/12763-pos-order-totals-order-creation-sync) to this branch to make sure we're testing & development with the latest order sync logic. However, there are some diffs in this PR from `trunk` because `feat/12763-pos-order-totals-order-creation-sync` might not be on the latest. Hopefully, https://github.com/woocommerce/woocommerce-ios/pull/13099 can be merged to have a clean state for this PR.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: Great point and clarification! Actually, this can't be easily implemented outside of the payment orchestrator because it has an internal token for this. I updated `cancelPayment` to call the one from the payment orchestrator in https://github.com/woocommerce/woocommerce-ios/pull/13100/commits/9c1c698edb56e0e74909ce4d42f57c95b7f7a671 to ensure it includes the wallet functionality recovery.

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift]: Josh implemented the better specialized PaymentOrchestrator approach in https://github.com/woocommerce/woocommerce-ios/pull/13100/commits/ecd2a868a56c5915108826f0b97eebfa13751259.

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:210]: Good question - from my testing, if we don't have the `onDisappear` trigger in the totals view, `onCancel` still gets triggered but [after syncing the order and starting payment collection](https://github.com/woocommerce/woocommerce-ios/blob/2a713c6cb2974da89118e98e6e5e20ac3e4c4f96/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentService.swift#L81-L82) instead of immediately after leaving the totals view when the task is replaced. I think canceling right away would be safer, WDYT?

**Comment by jaclync** [comment]
> The `.processing` case is particularly difficult, and that really just reflects that we can't always effectively cancel an in-progress payment. What do you think the right behaviour is in this case? We can't throw anything from the task's cancellation handler, unfortunately.

Yea this is a tough case, I think it's better to prevent leaving this screen (other than force quitting the app) so that the payment doesn't get interrupted (like those "Do not close or refresh the window" warnings after processing payment on websites 😄 ). We already have a task for this https://github.com/woocommerce/woocommerce-ios/issues/13115.

**Comment by jaclync** [comment]
@joshheald I ran out of time to test every scenario (I added checkmarks to the scenarios I've tested), but I updated the PR to fix the order total amount issue in PI p1718995930894699/1718983933.798199-slack-C070SJRA8DP and other updates. I think it's at a pretty good place at least for the first pass, would like to see what you think. Some followup changes unrelated to the cancelation work but noticed from testing:

- (In a separate PR) Hide the payment UI in the totals view until the order is synced and payment collection state is determined (auto-payment or not), otherwise the UI from the previous payment collection task could be shown like briefly having:

IMG_0387

- (In a separate PR) Show placeholder order totals text instead of incorrect amount like in the screenshot above for the $270

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
Comment (joshheald) [reply WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:210]: Yep, sounds good if that's what you've found 😊

File: Yosemite/Yosemite/Tools/POS/POSOrderService.swift
Comment (joshheald) [inline Yosemite/Yosemite/Tools/POS/POSOrderService.swift:138]: Is there any way you can think of that we could fix this underlying issue, so we can avoid upserting the order and all the potential issues it could cause?

**Review by joshheald — APPROVED** [review]
Thanks @jaclync – this works well.

## Test findings

### Issue with doubled-items order
One other major issue that we should look into: when testing, I once managed to get the order to be doubled up. 

The totals didn't add up – the subtotal and taxes matched the basket, which had 3 beanies – on my store that should have cost $60+$15 in taxes = $75. 

However, the total said $150, and that's what the shopper was charged (successfully.)

I'm pretty sure that this was a syncing/update issue from going backwards and forwards. When I viewed the final order, it had 2 line items for the same item, 3x Beanie.

If the sync had worked properly, there should only have been 1.

To make this safer, we could consider making a new order each time, as well as a new payment intent. I still don't love that for making lots of unused orders though.

### `auto-draft` orders showing in the existing app's order list.

I was able to see some unpaid `auto-draft` orders in the existing app's order list. I suspect some of that comes from the changes to the `POSOrderService` – the orders showed with notes showing that they had an initial status of `pending`, and were changed to `auto-draft`.

I don't think these should show up – I don't know why they are exactly, but perhaps the logic to hide them is actually on the server and now that we're upserting them to Storage, we're breaking that. I've added a question around whether the upsert is the only/best way to fix it...

### Pre-existing cancellation issue
I discovered a pre-existing issue with cancelling a reader connection, which impacts testing this change. The issue is with the underlying code, and can be recreated from the settings screen as well

1. Go to Menu > Payments > Manage Card Reader
2. Tap `Connect card reader`
3. Tap `Cancel` on the alert asking if you want to connect a particular reader
4. Tap `Connect card reader` again
5. Tap `Connect` on the reader connection alert
6. Observe that it stays stuck on the `Connecting to reader` screen forever

The issue predates POS – I have recreated it as far back as version 18.0 of the app, so it's probably existed for a very long time.

## Tests of the PR

### Critical flow

- [x] scanningForReaders(endSearch: () -> Void)
	- With a reader active nearby, the reader isn't auto-connected after leaving the totals screen to the product selector
	- As mentioned above, subsequent connections don't work, but this is unrelated to the change here.
- [x] connectingToReader
	- The reader still gets connected, but prevents the payment and shows invalidated payment in payment collection. 
	- The auto-payment succeeds after updating the cart
- [x] foundReader(name: String, connect: () -> Void, continueSearch: () -> Void, endSearch: () -> Void)
	- Reader isn't connected after leaving the totals screen to the product selector
	- As mentioned above, subsequent connections don't work, but this is unrelated to the change here.
- [x] showReaderList when multiple readers are found
	- Cancels as expected
	- As mentioned above, subsequent connections don't work, but this is unrelated to the change here.
- [x] preparingForPayment(cancelPayment: () -> Void)
	- The payment is not started, and a new one is made as expected after updating the cart
- [x] tapSwipeOrInsertCard(inputMethods: CardReaderInput, cancelPayment: () -> Void)
	- After leaving the totals screen in this state and then back, the reader can still transition to the ready state
	- In one case, the order contents were doubled! But I can't repro this... Perhaps we should consider also making a new order when we edit?
- [x] paymentSuccess(done: () -> Void)
	- I'm not really sure what cancellation looks like here. It shouldn't be possible to go back in this case (though it is right now.)
	- It shows an "Unable to process payment – this order is already paid" error, with a `Cancel Payment` button. The button resets the payment, but should probably do the same as `New transaction`. Otherwise, the user can press `Collect Payment` again and get the same error repeatedly.
- [x] paymentError(error: any Error, tryAgain: () -> Void, cancelPayment: () -> Void)
	- Recovers as expected
- [x] paymentErrorNonRetryable(error: any Error, cancelPayment: () -> Void)
	- Restarts the payment as expected, and shows the error again if needed.
- [x] processing
	- Really unpredictable. We should make it very difficult to attempt to cancel at this step, really.
	- Rarely, it leaves the reader waiting for a 30s timeout. But it's recoverable after that.
	- Fairly often, the payment will just complete.
- [x] displayReaderMessage(message: String)
	- Doing this at the `Remove card` reader message works as expected. It happens just before `processing`, so cancellation is still effective. This shouldn't be possible in the full implementation though.
- [x] validatingOrder(cancelPayment: () -> Void)
	- After leaving the totals screen in this state and then back, the reader can still transition to the ready state
	- Shows an error (from InvalidatablePaymentOrchestrator) which we can probably hide from the user, as it gets recovered correctly

### Test with a simulated reader

- [x] updateProgress(requiredUpdate: Bool, progress: Float, cancelUpdate: (() -> Void)?)
	- Cancels as expected, but similar connection failures as with other connection cancellations.
- [x] updateFailedLowBattery(batteryLevel: Double?, cancelUpdate: () -> Void)
	- Cancelling here shows the empty "Several readers found" list afterwards, on top of the cart screen. That's probably not what we want...
	- It's recoverable, showing the list with readers in it when you go back to the payment screen... but perhaps we should have reset it by now.
	- This behaviour matches tapping `Cancel` on the dialog
- [ ] updateFailed(tryAgain: () -> Void, cancelUpdate: () -> Void)
	- Couldn't test this with a real scenario
- [ ] updateFailedNonRetryable(cancelUpdate: () -> Void)
	- Couldn't test this with a real scenario

### Unhappy cases

- [ ] scanningFailed(error: Error, endSearch: () -> Void)
	- Couldn't test this with a real scenario
- [x] bluetoothRequired(error: Error, endSearch: () -> Void)
	- This error isn't thrown when it should be – separate issue, pre-existing. Seems like it's only done in the Manage Card Reader screen.
- [x] connectingFailed(error: Error, retrySearch: () -> Void, endSearch: () -> Void)
	- Going back during an ongoing connection attempt which would eventually fail didn't cancel it – that might not be possible though. However, it popped up the error message on top of the cart after going back. Repro: let reader be discovered, turn it off, then tap connect
	- If the error was already shown, it cancelled as expected
- [ ] connectingFailedNonRetryable(error: Error, endSearch: () -> Void)
	- Couldn't test this with a real scenario – after tapping try again, I couldn't get a repeatable error because of the previously mentioned cancellation/reconnection issue.
- [ ] connectingFailedUpdatePostalCode(retrySearch: () -> Void, endSearch: () -> Void)
	- Couldn't test this with a real scenario – seems like we only ever get the address one below.
- [x] connectingFailedUpdateAddress(wcSettingsAdminURL: URL, retrySearch: () -> Void, endSearch: () -> Void)
- [ ] connectingFailedChargeReader(retrySearch: () -> Void, endSearch: () -> Void)
	- Couldn't test this with a real scenario.

### Not supported

- [ ] selectSearchType(tapToPay: () -> Void, bluetooth: () -> Void, endSearch: () -> Void)
- [ ] cancelledOnReader

File: Yosemite/Yosemite/Tools/POS/POSOrderService.swift
Comment (jaclync) [reply Yosemite/Yosemite/Tools/POS/POSOrderService.swift:138]: The issue is from `CollectOrderPaymentUseCase` dispatching `OrderAction.retrieveOrder` to either load the order from remote (for the first time) or storage (after the first time): 

https://github.com/woocommerce/woocommerce-ios/blob/4b4d37e579477b02943453462fb13b5b86ee5c57/WooCommerce/Classes/ViewRelated/Orders/Collect%20Payments/CollectOrderPaymentUseCase.swift#L250

We could also rely on the `Order` passed to `CollectOrderPaymentUseCase` to check on the amount & paid status:

https://github.com/woocommerce/woocommerce-ios/blob/4b4d37e579477b02943453462fb13b5b86ee5c57/WooCommerce/Classes/ViewRelated/Orders/Collect%20Payments/CollectOrderPaymentUseCase.swift#L263-L269

Like by adding a parameter to `CollectOrderPaymentUseCase.collectPayment` like `shouldRetrieveOrderBeforeValidation: Bool` so that it's default to `true` for the existing app and `false` for POS (and maybe future use cases) to skip this Yosemite action and use the DI'ed order. This way, only a small amount of changes are in the existing app code and no changes are expected in the existing app flow. WDYT of this approach? I imagine there was a reason `OrderAction.retrieveOrder` is needed in `CollectOrderPaymentUseCase` in the first place.

**Comment by jaclync** [comment]
> Issue with doubled-items order

I was able to reproduce this issue with incorrect order total (screenshot below), my guess is from the API behavior syncing order before the previous synced order comes back. More details in the technical details in p91TBi-btJ-p2#comment-12385:
Order items in each order update API request are applied in addition to the latest order items on the backend. One way to ensure the items are correct is to wait for the previous order sync response before making the next one. The tradeoff is performance as it can take longer to sync the order, plus handling edge cases like if the previous order sync fails and it probably requires retries. We could also explore other options to reset order items in the order update API on the client side, maybe there are some missed parameters.

Creating a new order is also an option, and it would be the safest option as it guarantees a new PI as Stripe suggested, but with the tradeoff of generating unused orders in the database similar to the discussion in pdfdoF-54W-p2.

I plan to take a look at any other ways to reset order items in the order update API today.


Notice that the subtotal and total are inconsistent.


> auto-draft orders showing in the existing app's order list.

Yea, I believe these are a side effect from upserting the orders to storage. After orders tab is synced, they disappear fro the order list. `POSOrderService` creates the order with auto-draft status (`updateOrderStatus(posOrder:status:)` isn’t called at the moment, the status update is triggered by Collect Payment use case I think).

Discussion about alternatives of upserting orders to storage in https://github.com/woocommerce/woocommerce-ios/pull/13100#discussion_r1650712894.

If we keep upserting the orders to storage, we can see how order creation prevents auto-draft orders from appearing in the orders tab - maybe there’s some special logic there.

> Pre-existing cancellation issue

Thanks for creating the issue!

File: Yosemite/Yosemite/Tools/POS/POSOrderService.swift
Comment (joshheald) [reply Yosemite/Yosemite/Tools/POS/POSOrderService.swift:138]: The call during the payment flow is intended to make sure that we're working with the most up to date version of the order possible, in case it's been paid from some other source.

That's more relevant to the order list... however, if it's getting the order from _storage_ the second time, it's not really doing the job properly. We check because we want to make sure we're not using a copy of an order that's a few days out of date from Storage.

I don't think we should use the DI'ed Order, but we could change the app call to always get it from the network.

**Comment by jaclync** [comment]
Thanks a lot for the comprehensive & nicely documented testing in https://github.com/woocommerce/woocommerce-ios/pull/13100#pullrequestreview-2135255822 @joshheald! I've also just spent some time testing as many cases as I could, I'm going to merge this first and work on checking the order update API for the incorrect order total issue https://github.com/woocommerce/woocommerce-ios/issues/13127.

> paymentSuccess

This should be made impossible when we make the totals UI fullscreen:

Screenshot 2024-06-24 at 9 29 00 AM


> paymentErrorNonRetryable(error: any Error, cancelPayment: () -> Void)

How did you trigger a non-retryable error? I tried with turning device to offline after the ready state / before tapping card, and tapping a real card. Both resulted in retryable errors.

**Comment by joshheald** [comment]
> How did you trigger a non-retryable error? I tried with turning device to offline after the ready state / before tapping card, and tapping a real card. Both resulted in retryable errors.

Most of the time, if you get the same error during a retry, it will be transformed to a non-retryable error.

That's pretty naive, and we might want to change it, but that's how it goes at the moment I think.

**Comment by jaclync** [comment]
> Most of the time, if you get the same error during a retry, it will be transformed to a non-retryable error.
> That's pretty naive, and we might want to change it, but that's how it goes at the moment I think.

Thanks this works (to reproduce the non-retryable error), this is the last point you raised earlier in p1718735478831209/1718734994.548519-slack-C070SJRA8DP. I added a [draft task](https://github.com/orgs/woocommerce/projects/247/views/6?sliceBy%5Bvalue%5D=iOS+%F0%9F%8D%8E&pane=issue&itemId=68562328) to "To consider after MVP" column since it involves updating the existing app behavior.
I wonder in POS if we could offer a retry in this case to recreate the Payment Intent, if it's a payment error (like how going back to the product selector and back could fix it).

**Comment by joshheald** [comment]
> I was able to reproduce this issue with incorrect order total (screenshot below), my guess is from the API behavior syncing order before the previous synced order comes back. More details in the technical details in p91TBi-btJ-p2#comment-12385: Order items in each order update API request are applied in addition to the latest order items on the backend. One way to ensure the items are correct is to wait for the previous order sync response before making the next one. The tradeoff is performance as it can take longer to sync the order, plus handling edge cases like if the previous order sync fails and it probably requires retries. We could also explore other options to reset order items in the order update API on the client side, maybe there are some missed parameters.
> 
> Creating a new order is also an option, and it would be the safest option as it guarantees a new PI as Stripe suggested, but with the tradeoff of generating unused orders in the database similar to the discussion in pdfdoF-54W-p2.
> 
> I plan to take a look at any other ways to reset order items in the order update API today.

Nice one! I'll leave this to you, and focus on other unhappy path checks with the remainder of my day and tomorrow :)

As mentioned in the other thread – consider changing the original code to match the original intent: it should always get the latest copy of the order from the network, not storage. The current implementation just uses that order, but an improvement would be if it threw an error for a difference in the order totals between the one it was passed in and the one it found on the network.

**Comment by joshheald** [comment]
> I wonder in POS if we could offer a retry in this case to recreate the Payment Intent, if it's a payment error (like how going back to the product selector and back could fix it).

I think it depends on the error really – after MVP perhaps we can make our handling a bit more sophisticated for both the existing app and the POS.

The example I mentioned in slack (tapping a card and having it declined, then trying again with the same one) should be retried with the _same_ Payment Intent ideally, because nothing about the PI is wrong and it makes it less likely that we'll end up with double charges.

File: Yosemite/Yosemite/Tools/POS/POSOrderService.swift
Comment (jaclync) [reply Yosemite/Yosemite/Tools/POS/POSOrderService.swift:138]: > The call during the payment flow is intended to make sure that we're working with the most up to date version of the order possible, in case it's been paid from some other source.

In this case, always loading the order remotely feels like the expected behavior. Two ways I can think of:
- Add a parameter `OrderAction.retrieveOrder` like `alwaysFromRemote: Bool` to always load the order from the network for the payment collection use case
- Create a separate action like `OrderAction.retrieveOrderRemotely` to always sync the order from the network
- As a performance optimization that `OrderAction.retrieveOrder` has internally, create a separate `OrderAction` like `loadOrderForPaymentValidation` that only loads the order's `total` and `datePaid` remotely which are the 2 fields used for checking if an order is valid for payment

Because this requires changing the existing app behavior, I'll create an issue after your confirmation.

Comment (joshheald) [reply Yosemite/Yosemite/Tools/POS/POSOrderService.swift:138]: > As a performance optimization that OrderAction.retrieveOrder has internally, create a separate OrderAction like loadOrderForPaymentValidation that only loads the order's total and datePaid remotely which are the 2 fields used for checking if an order is valid for payment

This one sounds nice to me, and even better it might be `checkPaymentStatus(of: Order) -> PaymentStatus`. That will need more changes to the logic though, so let's leave the optimisation for the future. There are lots of things which could be better about our decisions on whether orders are paid or not!

> Create a separate action like OrderAction.retrieveOrderRemotely to always sync the order from the network

Let's do this one 😊 it's clear, simple, and doesn't involve changing the payments code significantly.

Comment (jaclync) [reply Yosemite/Yosemite/Tools/POS/POSOrderService.swift:138]: Thanks for the quick confirmation! I added a subtask to replace the upsert logic with this new action in https://github.com/woocommerce/woocommerce-ios/issues/12869, will work on this after checking the order update API (can also hand this off to Bozidar after time-boxing the effort since it's more on order sync than payment).

**Pull request merged**
